packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts

```diff
@@ -55,64 +66,77 @@ function mixPaths(
-                const fill = allFills[i];
-                const nextPath = [...path, fill];
+            for (let i = 0; i < remainingFills.length && steps < _maxSteps; ++i) {
+                const fill = remainingFills[i];
                 // Only walk valid paths.
-                if (!isValidPath(nextPath, true)) {
+                if (!isValidNextPathFill(path, flags, fill)) {
                     continue;
                 }
                 // Remove this fill from the next list of candidate fills.
-                const nextAllFills = allFills.slice();
-                nextAllFills.splice(i, 1);
+                const nextRemainingFills = remainingFills.slice();
+                nextRemainingFills.splice(i, 1);
                 // Recurse.
                 _walk(
-                    nextPath,
+                    [...path, fill],
                     input.plus(BigNumber.min(remainingInput, fill.input)),
                     output.plus(
                         // Clip the output of the next fill to the remaining
                         // input.
                         clipFillAdjustedOutput(fill, remainingInput),
                     ),
-                    nextAllFills,
+                    flags | fill.flags,
+                    nextRemainingFills,
                 );
             }
         }
     };
-    const allPaths = [...pathA, ...pathB];
-    const sources = allPaths.filter(f => f.index === 0).map(f => f.sourcePathId);
+    const allFills = [...pathA, ...pathB];
+    const sources = allFills.filter(f => f.index === 0).map(f => f.sourcePathId);
     const rateBySource = Object.assign(
         {},
         ...sources.map(s => ({
-            [s]: getPathAdjustedRate(side, allPaths.filter(f => f.sourcePathId === s), targetInput),
+            [s]: getPathAdjustedRate(side, allFills.filter(f => f.sourcePathId === s), targetInput),
         })),
     );
-    _walk(
-        [],
-        ZERO_AMOUNT,
-        ZERO_AMOUNT,
-        // Sort subpaths by rate and keep fills contiguous to improve our
-        // chances of walking ideal, valid paths first.
-        allPaths.sort((a, b) => {
-            if (a.sourcePathId !== b.sourcePathId) {
-                return rateBySource[b.sourcePathId].comparedTo(rateBySource[a.sourcePathId]);
-            }
-            return a.index - b.index;
-        }),
-    );
+    // Sort subpaths by rate and keep fills contiguous to improve our
+    // chances of walking ideal, valid paths first.
+    const sortedFills = allFills.sort((a, b) => {
+        if (a.sourcePathId !== b.sourcePathId) {
+            return rateBySource[b.sourcePathId].comparedTo(rateBySource[a.sourcePathId]);
+        }
+        return a.index - b.index;
+    });
+    _walk([], ZERO_AMOUNT, ZERO_AMOUNT, 0, sortedFills);
+    if (!isValidPath(bestPath)) {
+        throw new Error('nooope');
+    }
     return bestPath;
 }
 
+function isValidNextPathFill(path: Fill[], pathFlags: number, fill: Fill): boolean {
```

Comment:
We try to compute everything incrementally now so we don't need to continuously go over the path at every step.